### PR TITLE
Refactor stretch muting during NMO correction

### PR DESCRIPTION
**Pull request by alexanderkuvaev**
Stretch muting was reworked to match conventional software: an example of a gather being muted with the same parameters on `master` and the current branch is shown below:
mute_diff

Moreover, NMO correction and spectrum calculation functions were rewritten. Most noticeable changes include:
* NMO speed up by 20% for small gathers (~150 traces) and up to 50% for large gathers (~7k traces),
* Vertical velocity spectrum speed up by 35% for small gathers and up to 220% for large gathers for the same number of velocities,
* Residual velocity spectrum speed up by 300% for small gathers and up to 540% for large gathers for the same number of velocities - mostly due to a bug in the implementation here,
* Now both NMO and spectrum calculation functions don't require a gather to be sorted by offset,
* Now `apply_nmo` always performs muting of areas where time reversal occurred and thus `mute_crossover` flag was removed.

### CODE REVIEW

File: seismicpro/gather/utils/correction.py [hunk 1]
```diff
@@ -127,45 +95,60 @@ def apply_nmo(gather_data, times, offsets, stacking_velocities, sample_interval,
...
+    # Calculate hodograph times and an offset when time reversal or excessive stretching occurs
+    # for each zero-offset time
+    hodograph_times = np.empty((len(times), len(offsets)), dtype=np.float32)
+    max_offsets = np.empty(len(times), dtype=np.float32)
+    for i in prange(len(times)):
+        hodograph_times[i] = np.sqrt(times[i]**2 + (offsets / velocities[i])**2)
+        corrected_t0 = times[i] - offsets**2 * velocities_grad[i] / velocities[i]**3
+        stretch_mask = (corrected_t0 <= 0) | (hodograph_times[i] > (1 + max_stretch_factor) * corrected_t0)
+        muted_offsets = offsets[stretch_mask]
+        max_offsets[i] = muted_offsets.min() if muted_offsets.size > 0 else np.inf
```
Comment (nikita-klsh) [inline seismicpro/gather/utils/correction.py:137]: should we consider `max_offset =  offsets[~stretch_mask].max()` here,
and non-strict equality in `get_hodograph`: `if offsets[i] <= max_offset`

**Review by nikita-klsh — APPROVED** [review]

File: seismicpro/gather/utils/correction.py [hunk 2]
```diff
@@ -127,45 +95,60 @@ def apply_nmo(gather_data, times, offsets, stacking_velocities, sample_interval,
...
     delay : float
         Delay recording time of seismic traces. Measured in milliseconds.
-    mute_crossover: bool, optional, defaults to False
-        Whether to mute areas where time reversal occurred after NMO correction.
+    times : 1d np.ndarray
+        Zero-offset traveltimes to perform correction for. Measured in milliseconds.
+    velocities : 1d np.ndarray
+        Stacking velocities for each zero-offset traveltime. Must match the shape of `times`.
+        Measured in meters/milliseconds.
+    velocities_grad : 1d np.ndarray
```
Comment (anton-br) [inline seismicpro/gather/utils/correction.py:110]: From the docs it is not clear that any muting is applied. You can infer this fact from this and `max_stretch_factor` params. And if `max_stretch_factor` is somehow understandable but this one confuses a lot.
Consider adding a note about mute and how is it done in the main part of the doc.

File: seismicpro/gather/utils/correction.py (hunk 1 above)
Comment (alexanderkuvaev) [reply seismicpro/gather/utils/correction.py:137]: Unfortunately no, you need to get the first stretch event and mute everything after it: generally a non-stretched sample may follow a stretched one for early times and high velocity gradient.

**Review by anton-br — APPROVED** [review]

File: seismicpro/gather/utils/correction.py [hunk 3]
```diff
@@ -127,45 +95,60 @@ def apply_nmo(gather_data, times, offsets, stacking_velocities, sample_interval,
...
-        get_hodograph(gather_data, offsets, hodograph_times[i], sample_interval, delay, fill_value=fill_value,
-                      max_offset=max_offsets[i], out=corrected_gather_data[:, i])
+    # Calculate hodograph times and an offset when time reversal or excessive stretching occurs
+    # for each zero-offset time
+    hodograph_times = np.empty((len(times), len(offsets)), dtype=np.float32)
+    max_offsets = np.empty(len(times), dtype=np.float32)
+    for i in prange(len(times)):
+        hodograph_times[i] = np.sqrt(times[i]**2 + (offsets / velocities[i])**2)
+        corrected_t0 = times[i] - offsets**2 * velocities_grad[i] / velocities[i]**3
+        stretch_mask = (corrected_t0 <= 0) | (hodograph_times[i] > (1 + max_stretch_factor) * corrected_t0)
```
[the remark this answers is not in the document]
Comment (alexanderkuvaev) [reply seismicpro/gather/utils/correction.py]: Split into two masks

File: seismicpro/gather/utils/correction.py (hunk 2 above)
Comment (alexanderkuvaev) [reply seismicpro/gather/utils/correction.py:115]: Clarified a bit

**Pull request merged**